internal/sync2/tee.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package sync2
+
+import (
+	"io"
+	"io/ioutil"
+	"sync"
+	"sync/atomic"
+)
+
+type tee struct {
+	buffer ReadAtWriteAtCloser
+	open   *int64
+
+	mu      sync.Mutex
+	nodata  sync.Cond
+	newdata sync.Cond
+
+	maxRequired int64
```

Comment:
maybe `readtail`, `readLimit` or similar?

Reply:
It stores the maximum byte number requested by any of the readers so far. The writer checks it to see if it is necessary to add more data to the buffer.

So maybe a name like `readHead` instead of `readTail`. Or `maxReadHead` to denote that it is about the read head that is furthest ahead.

Reply:
`maxRead`?

Reply:
Deal!